### PR TITLE
Create a text validator/formatter mechanism

### PR DESCRIPTION
**Pull request by xster**
#9511

Add a text formatter interface used by EditingText. Provide some default implementations.

### CODE REVIEW

**Comment by xster** [comment]
Soliciting some feedback before I put in all the tests

File: packages/flutter/lib/src/services/text_editing.dart [hunk 1]
```diff
@@ -195,4 +195,15 @@ class TextSelection extends TextRange {
     affinity.hashCode,
     isDirectional.hashCode
   );
+
+  /// Creates a new [TextSelection] based on the current selection, with the
+  /// provided parameters overridden.
+  TextSelection copyWith({int baseOffset, int extentOffset}) {
```
Comment (Hixie) [inline packages/flutter/lib/src/services/text_editing.dart]: why only those two instead of all four?

nit: spaces around named argument declarations

Comment (xster) [inline packages/flutter/lib/src/services/text_editing.dart]: it was arbitrary :) changed

File: packages/flutter/lib/src/widgets/text_formatter.dart [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  );
+}
+
+/// A [TextInputFormatter] that prevents the insertion of blacklisted.
+/// 
+/// Instances of blacklisted characters found in the new [TextEditingValue]s 
+/// will be replaced with the [replacementString] which defaults to ``.
+/// 
+/// It attempts to preserve the existing [TextEditingValue.selection] to 
+/// sensible values.
```
Comment (cbracken) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: Maybe: 
```
/// This formatter attempts to preserve the [TextEditingValue.selection].
```

That said, as-is, I'm not convinced this adds much. It feels like we're saying 'This formatter behaves how you'd expect it to.' Perhaps more directly state that this formatter will adjust the selection in the case that blacklisted characters were removed?

File: packages/flutter/lib/src/widgets/text_formatter.dart [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+          extentOffset: beforeSelection.length + inSelection.length,
+        ),
+        composing: TextRange.empty,
+      );
+    }
+  }
+
+  /// A [BlacklistingTextInputFormatter] that forces input to be a single line.
+  static final BlacklistingTextInputFormatter singleLineFormatter
+      = new BlacklistingTextInputFormatter(new RegExp(r'\n'));
```
Comment (cbracken) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: Is this correct? 

During a copy-paste, I'd assume a replacementString where every contiguous string of newlines is collapsed to a single `0x20` space. 

During text entry, _ideally_ I'd want `\n` to be replaced with the empty string, but in single-line text fields, I suspect we'll also set a keyboard that only supplies a `Done` key so using a space is probably a non-issue.

**Review by cbracken — CHANGES_REQUESTED** [review]

File: packages/flutter/lib/src/widgets/text_formatter.dart (hunk 3 above)
Comment (xster) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: Not sure I fully caught what you meant. Can you add an example? Or did you want r'\n+'? 

if the user pastes:
```


```
with a replacement string of `' '` (1 space), it outputting `'   '` (3 spaces) feels more predictable.

Comment (cbracken) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: Corrected my original statement. Should be a straight-up 1:1 replacement of `\n` with a space.

File: packages/flutter/lib/src/widgets/text_formatter.dart [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+/// sensible values. 
+/// 
+/// See also:
+///  
+///  * [TextInputFormatter].
+class BlacklistingTextInputFormatter extends TextInputFormatter {
+  BlacklistingTextInputFormatter(
+    this.blacklistedPattern, 
+    {this.replacementString: ''}) : 
+    assert(blacklistedPattern != null);
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: nit: the formatting here isn't consistent with the style we normally use

File: packages/flutter/lib/src/widgets/text_formatter.dart [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+  /// A [Pattern] to match and replace incoming [TextEditingValue]s.
+  final Pattern blacklistedPattern;
+  /// String used to replace found patterns.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: nit: newline between the members

File: packages/flutter/lib/src/widgets/text_formatter.dart [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+  final String replacementString;
+
+  @override
+  TextEditingValue formatEditUpdate(
+    TextEditingValue oldValue, 
+    TextEditingValue newValue
+  ) {
+    // Don't apply formatting to text still under composition.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: can we guarantee that you can't compose something with a newline?

File: packages/flutter/lib/src/widgets/text_formatter.dart (hunk 3 above)
Comment (xster) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: I think that's true no? If there's 20 \n in whichever placement, replaceAll will replace them all with 20 spaces right?

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 7]
```diff
@@ -133,7 +134,13 @@ class EditableText extends StatefulWidget {
     this.keyboardType,
     this.onChanged,
     this.onSubmitted,
-  }) : super(key: key) {
+    List<TextInputFormatter> inputFormatters,
+  }) : inputFormatters = maxLines == 1 
+           ? ((inputFormatters ?? <TextInputFormatter>[])
+                 ..add(BlacklistingTextInputFormatter.singleLineFormatter)
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/editable_text.dart]: mutating the list is sketchy. if someone creates one then passes it in each time, it'll get progressively longer.
if someone passes in a const list, it'll crash.

File: packages/flutter/lib/src/widgets/text_formatter.dart (hunk 6 above)
Comment (xster) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: Can't but I think it's a bit too edge case-y to worry about. My intent is just handle cases where if someone blacklists latin characters and starts typing chinese, there should still be latin characters in the composition interim. If the user presses send/save/wtv right at that moment, I think we should 'commit' the composition first (which triggers the formatting) and then pass it out.

Comment (Hixie) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: I think when you're in a single-line text field, we need to be able to guarantee that you cannot ever enter a newline.

File: packages/flutter/lib/src/widgets/text_formatter.dart (hunk 3 above)
Comment (cbracken) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: Exactly - yep.

File: packages/flutter/lib/src/widgets/text_formatter.dart [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+  /// 
+  /// You can override the resulting text based on the previous text value and
+  /// the incoming new text value.
+  TextEditingValue formatEditUpdate(
+    TextEditingValue oldValue, 
+    TextEditingValue newValue
+  );
+}
+
+/// A [TextInputFormatter] that prevents the insertion of blacklisted.
```
Comment (xster) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: done

File: packages/flutter/lib/src/widgets/text_formatter.dart (hunk 2 above)
Comment (xster) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: reworded a bit

File: packages/flutter/lib/src/widgets/text_formatter.dart (hunk 4 above)
Comment (xster) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: changed

File: packages/flutter/lib/src/widgets/text_formatter.dart (hunk 5 above)
Comment (xster) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: done

File: packages/flutter/lib/src/widgets/text_formatter.dart (hunk 6 above)
Comment (xster) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: I was pedantic with the wording. But pragmatically, I don't think it's possible to enter a newline while composing.

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 7 above)
Comment (xster) [inline packages/flutter/lib/src/widgets/editable_text.dart]: Ya, good call, that was a disaster... Changed

**Comment by xster** [comment]
I expanded this PR a bit to avoid diffbase gymnastics. 
- Added a WhitelistingTextInputFormatter
- Use a digits only whitelisting formatter in the gallery demo
- Add a phone number formatter in the gallery demo

Now fixes #9511. Fixes #9340. And fixes a bug in the gallery on iOS that prevents me from testing #9200.

File: examples/flutter_gallery/lib/demo/material/text_form_field_demo.dart [hunk 9]
```diff
@@ -59,9 +60,9 @@ class TextFormFieldDemoState extends State<TextFormFieldDemo> {
 
   String _validatePhoneNumber(String value) {
     _formWasEdited = true;
-    final RegExp phoneExp = new RegExp(r'^\d\d\d-\d\d\d\-\d\d\d\d$');
+    final RegExp phoneExp = new RegExp(r'^\(\d\d\d\) \d\d\d\-\d\d\d\d$');
     if (!phoneExp.hasMatch(value))
-      return '###-###-#### - Please enter a valid phone number.';
+      return '(###) ###-#### - Please enter a valid phone number.';
```
Comment (Hixie) [inline examples/flutter_gallery/lib/demo/material/text_form_field_demo.dart]: US phone number

File: packages/flutter/lib/src/widgets/text_formatter.dart [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+/// fall at with the removed characters.
+/// 
+/// See also:
+///  
+///  * [TextInputFormatter].
+///  * [WhitelistingTextInputFormatter].
+class BlacklistingTextInputFormatter extends TextInputFormatter {
+  BlacklistingTextInputFormatter(
+    this.blacklistedPattern, 
+    {this.replacementString: ''}
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: put the `{` on the previous line and the `}` on the next line to match similar code elsewhere in the codebase

File: packages/flutter/lib/src/widgets/text_formatter.dart [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+  TextEditingValue formatEditUpdate(
+    TextEditingValue oldValue, // unused.
+    TextEditingValue newValue,
+  ) {
+    // Don't apply formatting to text still under composition.
+    if (!newValue.composing.isCollapsed)
+      return newValue;
+    return _selectionAwareTextManipulation(
+      newValue, 
+      (String substring) =>
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: nit: don't use => if you have a line break

File: packages/flutter/lib/src/widgets/text_formatter.dart [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+  TextEditingValue formatEditUpdate(
+    TextEditingValue oldValue, // unused.
+    TextEditingValue newValue,
+  ) {
+    // Don't apply formatting to text still under composition.
+    if (!newValue.composing.isCollapsed)
+      return newValue;
+    return _selectionAwareTextManipulation(
+      newValue, 
+      (String substring) => whitelistedPattern
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: same here

File: packages/flutter/lib/src/widgets/text_formatter.dart (hunk 6 above)
Comment (Hixie) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: How about if the code passes in a newline in the original string? Or if someone writes a custom keyboard that does it?

File: examples/flutter_gallery/lib/demo/material/text_form_field_demo.dart (hunk 9 above)
Comment (xster) [inline examples/flutter_gallery/lib/demo/material/text_form_field_demo.dart]: Done

File: packages/flutter/lib/src/widgets/text_formatter.dart (hunk 10 above)
Comment (xster) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: Done

File: packages/flutter/lib/src/widgets/text_formatter.dart (hunk 11 above)
Comment (xster) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: Done

File: packages/flutter/lib/src/widgets/text_formatter.dart (hunk 12 above)
Comment (xster) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: Done

**Comment by xster** [comment]
Added tests

File: packages/flutter/lib/src/widgets/text_formatter.dart (hunk 6 above)
Comment (xster) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: what did you mean the code passes in? Did you mean if the user passes in a second formatter that breaks the first one?

Either way, added handling to format text under composition

Comment (Hixie) [inline packages/flutter/lib/src/widgets/text_formatter.dart]: I meant if someone writes Flutter code that sets the initial value of the editable text field to `'a\nb'`.

**Comment by xster** [comment]
Gentle ping

**Comment by Hixie** [comment]
LGTM but @cbracken should rereview since he marked it as X.

**Review by cbracken — APPROVED** [review]
lgtm

Since `TextInputFormatter` & co. aren't strictly dependent on anything in `widgets`, consider moving those to `services`. On the off chance anyone else decided to implement their own `widgets` alternative, they could still deal with the existing validators.

**Comment by xster** [comment]
oh ya, that's a really good point. Since the TextSelection stuff is already in services. Done

**Pull request merged**